x-pack/plugins/security/server/anonymous_access/anonymous_access_service.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import type { Request } from '@hapi/hapi';
+import { IBasePath, KibanaRequest, Logger } from '../../../../../src/core/server';
+import { addSpaceIdToPath } from '../../../spaces/common';
+import type { SpacesServiceStart } from '../../../spaces/server';
+import { AUTH_PROVIDER_HINT_QUERY_STRING_PARAMETER } from '../../common/constants';
+import { AnonymousAuthenticationProvider } from '../authentication';
+import type { AuthorizationServiceSetup } from '../authorization';
+import type { ConfigType } from '../config';
+
+export interface AnonymousAccessServiceSetup {
+  readonly isAnonymousAccessEnabled: boolean;
+}
+
+export interface AnonymousAccessServiceStart {
+  readonly isAnonymousAccessEnabled: boolean;
+  // We cannot use `ReadonlyMap` just yet: https://github.com/microsoft/TypeScript/issues/16840
+  readonly accessURLParameters: Readonly<Map<string, string>> | null;
+  isSavedObjectTypeAccessibleAnonymously: (
+    request: KibanaRequest,
+    savedObjectType: string
+  ) => Promise<boolean>;
+}
+
+interface AnonymousAccessServiceStartParams {
+  authz: Pick<AuthorizationServiceSetup, 'actions' | 'checkSavedObjectsPrivilegesWithRequest'>;
+  basePath: IBasePath;
+  spaces?: SpacesServiceStart;
+}
+
+/**
+ * Service that manages various aspects of the anonymous access.
+ */
+export class AnonymousAccessService {
+  constructor(private readonly logger: Logger, private readonly getConfig: () => ConfigType) {}
+
+  setup(): AnonymousAccessServiceSetup {
+    return {
+      isAnonymousAccessEnabled: this.isAnonymousAccessEnabled(),
+    };
+  }
+
+  start({
+    authz,
+    basePath,
+    spaces,
+  }: AnonymousAccessServiceStartParams): AnonymousAccessServiceStart {
+    const config = this.getConfig();
+    const anonymousProvider = config.authc.sortedProviders.find(
+      ({ type }) => type === AnonymousAuthenticationProvider.type
+    );
+    // We don't need to add any special parameters to the URL if any of the following is true:
+    // * anonymous provider isn't enabled
+    // * anonymous provider is enabled, but it's a default authentication mechanism
+    const anonymousIsDefault =
+      !config.authc.selector.enabled && anonymousProvider === config.authc.sortedProviders[0];
+
+    return {
+      isAnonymousAccessEnabled: this.isAnonymousAccessEnabled(),
+      accessURLParameters:
+        anonymousProvider && !anonymousIsDefault
```

Comment:
**question** is there harm in supplying the access url parameters if anonymous is the default provider? If an administrator changes the auth provider config, then they'd potentially invalidate any previously generated URLs if the anonymous provider suddenly isn't the default anymore.

Reply:
> question is there harm in supplying the access url parameters if anonymous is the default provider?

Well, the base assumption here (we'll need to verify) that anonymous access will be the only one or default one in the majority of the setups. And the idea is that Share plugin shouldn't show `Public URL` option during sharing/embedding  if there is nothing special to be added to the URL that would make UI less loaded and confusing.

> If an administrator changes the auth provider config, then they'd potentially invalidate any previously generated URLs if the anonymous provider suddenly isn't the default anymore.

Actually it feels reasonable to me to show login selector if the URL was shared without `Public URL` (was hidden when anonymous was default, so user didn't indicate that this URL should use anonymous access specifically) in case anonymous becomes non-default.

Do you see any issues here?

Reply:
> Well, the base assumption here (we'll need to verify) that anonymous access will be the only one or default one in the majority of the setups. And the idea is that Share plugin shouldn't show Public URL option during sharing/embedding if there is nothing special to be added to the URL that would make UI less loaded and confusing.

I'd personally be surprised if anonymous access was commonly the only provider, but I take your point here.

> Actually it feels reasonable to me to show login selector if the URL was shared without Public URL (was hidden when anonymous was default, so user didn't indicate that this URL should use anonymous access specifically) in case anonymous becomes non-default.

I'm torn on this. From a security perspective, it makes sense to fail closed if the auth config changes in a way that alters the default provider. On the other hand, the administrator has to take explicit action to make the anonymous provider the default/only provider, so you could argue that generating these share links is intentionally making them available anonymously.

Let's keep it the way you have it for now. This is something we can easily change if we decide to take a different approach